pkg/api/types/foo.go

```diff
@@ -0,0 +1,16 @@
+package types
+
+import uuid "github.com/satori/go.uuid"
+
+type FooReads struct {
+	Name       *string `json:"name"`
+	OtherField *int    `json:"otherField"`
+	Valid      *bool   `json:"valid"`
+}
+
+type Foo struct {
```

Comment:
If you want to have two types for this, you could embed FooReads into Foo
```
type Foo struct {
  FooReads
  ID uuid.UUID
}
```
then you can reuse the same validation logic for both types

Reply:
@neb42 There might be good reasons using a pointer, but I would only use it if you have a good reason, e.g.need to differentiate between nil or default value. It's usually best to just use the default value and let the consuner handle it, but it might feel a bit strange in JS.

Alternatively, you could use the omitempty tag which would remove the property if it's set to a default value then use pointers if you can't differentiate e.g. on bool or a nested struct.

https://www.sohamkamani.com/golang/omitempty/

Might not be an issue for you, but using pointers will also lead to more memory usage, need to store both the pointer and the value, slower json encoding and it also leads to more complexity. So in general it should be avoided unless it's needed as it's seldom an optimization.

The same applies to small struct passed as arguments
https://github.com/golang/go/wiki/CodeReviewComments#pass-values.

Here are aom good rules for receivers and pointers https://github.com/golang/go/wiki/CodeReviewComments#pass-values

Go is fast at copying data so e.g. returning a pointer from a function is also something you should have good reasons to do as it's not really an optimization (lead to more complex code and data would probably end up on the heap and require GC instead of on the stack).

If you have a choice, try to optimize for simplicity and maintainability. Clear is better than clever.

Reply:
> If you have a choice, try to optimize for simplicity and maintainability. Clear is better than clever

100% agree and thanks for the links, I'll definitely give them a read through.

I think my issue is that default values lead to bad api design (maybe slightly too strong a claim). On the client side I find it strange that not sending a field will give it a default value rather than an error. I personally prefer a little more internal complexity for a cleaner interface.

I guess this is just solved with good validation rules that error if the default value is invalid, eg. a name can't be an empty string.

I think it also introduces an extra layer of complexity when a value could be null.

Reply:
The philosophy between Go and JavaScript is a bit different here. In old days we would just check for falsy in JS as it would also include the default values of false, "" and 0. Now I guess it's cleaner to use omitempty tag on the Go side and then just check for foo === undefined or foo ?? error();

Reply:
Regarding pointers, this is also an interesting article https://medium.com/@vCabbage/go-are-pointers-a-performance-optimization-a95840d3ef85

Reply:
Haha, I think javascript also has it's issues with falsy values.

I think checking for undefined on the response is fine (not my preference but I'm ok with it)

My issue is more on the request. I think if a field is missing it should return a 422 rather than give a default value.

Reply:
Aha I didn't understand this was for a request. Personally I would just check for empty value.  Or use a pointer like you did. A few other options as well https://stackoverflow.com/questions/19633763/unmarshaling-json-in-go-required-field

Reply:
Yeah I think my approach is to define a `reads` struct using pointers, allowing us to check for `nil` where appropriate. This then gets put into an internal struct (which I also use for the `writes`) using values.

Although if I refactored it to use:
```
type Foo struct {
  FooReads
  ID uuid.UUID
}
```
then this would still use pointers.

> Personally I would just check for empty value

I think this would mostly work, until it doesn't. For strings, I would typically have an empty string and null be the same so I can live with that. Booleans are fine as false. It's numbers having the default as 0 that causes an issue for me. Javascript also annoys me with 0's being falsy.

Reply:
Yea. SQL has nullable types, but you need to handle that yourself in JSON, or you could use https://github.com/guregu/null

Reply:
Yeah I did see that package a while ago. It may be worth trying out if this becomes an issue. I think I need to spend more time with go first though (my backend background is with node/scala/python).